src/DynamoCore/Core/Threading/AsyncTaskExtensions.cs

```diff
@@ -7,24 +7,101 @@
 
 namespace Dynamo.Core.Threading
 {
+    /// <summary>
+    ///     Tools for working productively with AsyncTask's
+    /// </summary>
     internal static class AsyncTaskExtensions
     {
+        /// <summary>
+        ///     Upon completion of the task, invoke the specified action
+        /// </summary>
+        /// <returns>An IDisposable representing the event subscription</returns>
         internal static IDisposable Then(this AsyncTask task, AsyncTaskCompletedHandler action)
         {
             task.Completed += action;
             return Disposable.Create(() => task.Completed -= action);
         }
 
+        /// <summary>
+        ///     Upon completion of the task, invoke the specified asynchronously in the specified
+        ///     SynchronizationContext
+        /// </summary>
+        /// <returns>An IDisposable representing the event subscription</returns>
         internal static IDisposable ThenPost(this AsyncTask task, AsyncTaskCompletedHandler action, SynchronizationContext context = null)
         {
             if (context == null) context = new SynchronizationContext(); // uses the default
             return task.Then((t) => context.Post((_) => action(task), null));
         }
 
+        /// <summary>
+        ///     Upon completion of the task, invoke the specified synchronously in the specified
+        ///     SynchronizationContext
+        /// </summary>
+        /// <returns>An IDisposable representing the event subscription</returns>
         internal static IDisposable ThenSend(this AsyncTask task, AsyncTaskCompletedHandler action, SynchronizationContext context = null)
         {
             if (context == null) context = new SynchronizationContext(); // uses the default
             return task.Then((t) => context.Send((_) => action(task), null));
         }
+
+        /// <summary>
+        ///     Await the completion of a collection of scheduled tasks.  The given action
+        ///     will only be executed after all events are complete.  The tasks must already
+        ///     be scheduled or this action will never be executed.
+        /// </summary>
+        /// <returns>An IDisposable representing all of the event subscription</returns>
+        internal static IDisposable AllComplete(this IEnumerable<AsyncTask> tasks, Action<IEnumerable<AsyncTask>> action)
+        {
+            // If the task list is empty, we immediately invoke the action
+            if (!tasks.Any())
+            {
+                action(tasks);
+            }
+
+            // We'll have to keep track of the event subscriptions for disposal
+            var cbs = new List<Tuple<AsyncTask, AsyncTaskCompletedHandler>>();
```

Comment:
I guess you don't like terse decs? `callbacks`

On Tue, May 12, 2015, 6:56 PM Luke Church <notifications@github.com> wrote:

> In src/DynamoCore/Core/Threading/AsyncTaskExtensions.cs
> <https://github.com/DynamoDS/Dynamo/pull/4424#discussion_r30189256>:
>
> > +        /// <summary>
> > +        ///     Await the completion of a collection of scheduled tasks.  The given action
> > +        ///     will only be executed after all events are complete.  The tasks must already
> > +        ///     be scheduled or this action will never be executed.
> > +        /// </summary>
> > +        /// <returns>An IDisposable representing all of the event subscription</returns>
> > +        internal static IDisposable AllComplete(this IEnumerable<AsyncTask> tasks, Action<IEnumerable<AsyncTask>> action)
> > +        {
> > +            // If the task list is empty, we immediately invoke the action
> > +            if (!tasks.Any())
> > +            {
> > +                action(tasks);
> > +            }
> > +
> > +            // We'll have to keep track of the event subscriptions for disposal
> > +            var cbs = new List<Tuple<AsyncTask, AsyncTaskCompletedHandler>>();
>
> cbs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DynamoDS/Dynamo/pull/4424/files#r30189256>.
>